### PR TITLE
For second-time build, let build_type be inferred from CMakeCache.txt.

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23331 Update doc about subsequent builds: options can be changed in build/CMakeCache.txt
* **#23323 For second-time build, let build_type be inferred from CMakeCache.txt.**

This leads to more consistent build outcome.

Differential Revision: [D16517621](https://our.internmc.facebook.com/intern/diff/D16517621)

### CODE REVIEW

**Comment by xuhdev** [comment]
get_cmake_cache_variables_from_file and convert_cmake_value_to_python_value are moved to module-level functions to ease the use of other modules.

File: tools/setup_helpers/cmake.py [hunk 1]
```diff
@@ -31,6 +31,57 @@ def _mkdir_p(d):
 USE_NINJA = (not check_negative_env_flag('USE_NINJA') and
              which('ninja') is not None)
 
+def convert_cmake_value_to_python_value(cmake_value, cmake_type):
```
Comment (ezyang) [inline tools/setup_helpers/cmake.py]: Any changes here?

**Comment by ezyang** [comment]
One question that I have, is that someone might reasonably run `python setup.py develop`, and then subsequently run `CMAKE_BUILD_TYPE=Debug python setup.py develop` to try to make a debug build. But the environment variable will be silently ignored until they force a re-cmake. The trouble is we don't have a clear distinction between "configure" and "build" phases.

In Cabal, we designed a pretty nice user pattern to handle this case. Here's how it worked:

* If you ran `cabal new-build`, options you passed to the script were ephemeral. If you changed the options, we would immediately rebuild respecting the new options. The build system, when interacted with via `new-build`, is completely stateless.
* If you ran `cabal new-configure`, options you passed would be persisted to a human readable local file. This file is now considered when you make subsequent invocations to `new-build`: it basically assigns defaults to options (but you could still override them with `new-build`). If you wanted to get rid of the file, you could just delete it; you could also edit it yourself.

Do you think something like that is appropriate here? I suppose I can land this first anyway, but I would at the very least be keen on warnings when we silently ignore user-given input.

**Comment by ezyang** [comment]
In your defense, the usage pattern I described probably has never worked. So if you say so, we can try landing this as is.

File: tools/setup_helpers/cmake.py (hunk 1 above)
Comment (xuhdev) [reply tools/setup_helpers/cmake.py]: I added a line number for error indication:

```diff
@@ -1,13 +1,12 @@
-@staticmethod
 def convert_cmake_value_to_python_value(cmake_value, cmake_type):
     r"""Convert a CMake value in a string form to a Python value.
 
     Arguments:
-      cmake_value (string): The CMake value in a string form (e.g., "ON", "OFF", "1").
-      cmake_type (string): The CMake type of :attr:`cmake_value`.
+        cmake_value (string): The CMake value in a string form (e.g., "ON", "OFF", "1").
+        cmake_type (string): The CMake type of :attr:`cmake_value`.
 
     Returns:
-      A Python value corresponding to :attr:`cmake_value` with type :attr:`cmake_type`.
+        A Python value corresponding to :attr:`cmake_value` with type :attr:`cmake_type`.
     """
 
     cmake_type = cmake_type.upper()
@@ -23,18 +22,17 @@
     else:  # Directly return the cmake_value.
         return cmake_value
 
-@staticmethod
-def _get_cmake_cache_variables(cmake_cache_file):
+def get_cmake_cache_variables_from_file(cmake_cache_file):
     r"""Gets values in CMakeCache.txt into a dictionary.
 
     Arguments:
-      cmake_cache_file: A CMakeCache.txt file object.
+        cmake_cache_file: A CMakeCache.txt file object.
     Returns:
-      dict: A ``dict`` containing the value of cached CMake variables.
+        dict: A ``dict`` containing the value of cached CMake variables.
     """
 
     results = dict()
-    for line in cmake_cache_file:
+    for i, line in enumerate(cmake_cache_file, 1):
         line = line.strip()
         if not line or line.startswith(('#', '//')):
             # Blank or comment line, skip
@@ -43,11 +41,11 @@
         # Space can also be part of variable name and value
         matched = re.match(r'(\S.*):\s*([a-zA-Z_-][a-zA-Z0-9_-]*)\s*=\s*(.*)', line)
         if matched is None:  # Illegal line
-            raise ValueError('Unexpected line in {}: {}'.format(repr(cmake_cache_file), line))
+            raise ValueError('Unexpected line {} in {}: {}'.format(i, repr(cmake_cache_file), line))
         variable, type_, value = matched.groups()
         if type_.upper() in ('INTERNAL', 'STATIC'):
             # CMake internal variable, do not touch
             continue
-        results[variable] = CMake.convert_cmake_value_to_python_value(value, type_)
+        results[variable] = convert_cmake_value_to_python_value(value, type_)
 
     return results
```

**Comment by xuhdev** [comment]
Does the system you described have a name? I'm not aware of such feature in CMake, which is probably a deal-breaker for using it in PyTorch.

But we actually somewhat have a configuration step, if we run with `--cmake-only`, while the default runs configuration and build altogether. Since we use env vars to configure options (if via `setup.py`), ideally, we can allow a build option via env vars re-adaptation if `--cmake` is switched on (Is this eventually what we want to have? I guess so.). But this not yet consistent (e.g., `CUDA_NVCC_EXECUTABLE`) and therefore I avoided documenting this, and only allow users to edit `build/CMakeCache.txt` for now.

But I can also make this build option adaptable when `--cmake` is switched on, either in this PR or in another PR.

**Comment by xuhdev** [comment]
In other words, we currently have

(plainly == no additional build options passed in to setup.py)

for clean tree:

- configuration only: `--cmake-only`
- configuration + build: invoke setup.py plainly

For rebuild:

- reconfiguration without rebuild: `--cmake-only --cmake` (unreliable), or edit CMakeCache.txt and run cmake directly
- rebuild without reconfiguration: invoke setup.py plainly
- reconfiguration + rebuild: `--cmake` (unreliable), or edit CMakeCache.txt and invoke setup.py plainly

I think the question is whether we wanna make the two "unreliable" spots reliable, so that we can enable `setup.py` lovers consistently use `setup.py` without explicitly switching to cmake, while still enjoying the benefit of viewing and editing all build options in one CMakeCache.txt.

**Comment by ezyang** [comment]
> I'm not aware of such feature in CMake, which is probably a deal-breaker for using it in PyTorch.

I think you are right that the corresponding feature in cmake is the difference between running `cmake` and running `make`. If you run make, no need to reproduce all of the configuration flags, and if necessary, cmake will reconfigure itself with the same set of variables.

> we can allow a build option via env vars re-adaptation if --cmake is switched on (Is this eventually what we want to have? I guess so.)

and

> I think the question is whether we wanna make the two "unreliable" spots reliable, so that we can enable setup.py lovers consistently use setup.py without explicitly switching to cmake, while still enjoying the benefit of viewing and editing all build options in one CMakeCache.txt.

I would definitely prefer if I could pass `--cmake`, change env vars and get the build system to pick up the changes. Most people don't know about `CMakeCache.txt` and they've been conditioned to just `python setup.py clean` when they want a config change and it doesn't seem to stick.

Of course, this is more work. So it doesn't have to happen now (or even in the near future). But I think it would be good to agree that, in an ideal world, this is how it would work.

I'm going to go ahead and land this.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**